Add support for implicit conversions to C++ types

jagerman opened the pull request:
This PR adds a new py::implicitly_cpp_convertible<From,To>() tells pybind() that From can be implicitly converted to To (in C++), and thus pybind11-registered From types can be used in methods that accept (C++) To instances.

(See issue #259)

In essence, this lets you make use of C++ implicit converters (typically via a 'From::operator To()' or a non-explicit 'To::To(const From&)' constructor) with pybind11 instances without needing to create an explicit pybind11 interface for the conversion.

As a simple example, consider the two classes:

```
    class A {
        ...
        void method(uint64_t val) { ... }
    }
    class B {
        ...
        operator uint64_t() { ... } // returns unique identifier
    }
```

In C++, you can call `a.method(22)` or `a.method(b)`.  Telling pybind11 about the conversion allows you to do the same in python.

Without the implicit conversion, you would have to either overload A.method to accept both uint64_t and B instances, or would have to add a method to B's interface that returns the id, then pass this method result to A's method.

wjakob commented:
Sorry, I must be missing something. How is this different from the existing py::implicitly_convertible?

jagerman commented:
py::implicitly_convertible only allows conversion to types declared by the user.  It also allows conversion between types aren't, at the C++ level, implicitly convertible.

This only allows conversion *from* types declared by the user, but allows conversion to any type that the C++ source type can be implicitly converted to.

wjakob commented:
A few thoughts: It's probably not a good idea to put implicitly_convertible checks into every type_caster (that's quite a lot of places! I am concerned about code generation bloat + maintainability issues).

As far as other places go, the tuple caster comes to mind (any function call will pass through that to convert the tuple of function arguments).

Also, it seems somehow inelegant to me to have two kinds of implicitly_convertible statements. It would be cleaner to have one unified declaration that subsumes both cases.

jagerman commented:
> It's probably not a good idea to put implicitly_convertible checks into every type_caster (that's quite a lot of places! I am concerned about code generation bloat + maintainability issues).

Yes, moving into the tuple caster seems like a better approach.  I'll do that.

> Also, it seems somehow inelegant to me to have two kinds of implicitly_convertible statements. It would be cleaner to have one unified declaration that subsumes both cases.

I started out with updating implicitly_convertible, but I ended up doing two cases because the meanings are a bit different.  py::implicitly_convertible<A,B> is essentially saying that when you write (in python) `func(a)` we'll translate that to `func(B(a))` if func accepts B instances but not A instances.  B(A) must be a registered constructor and will be used for the conversion.

The cpp one is a little different because it allows implicit conversion behind-the-scenes in almost any way that C++ allows it, which can be via a constructor in B, but can also be done through an `operator B()` in class A.  What that means in practice is that py::implicitly_cpp_convertible<A,B> allows me to write `func(a)` in python even if I haven't exposed a `B(A)` constructor in Python, as long as the constructor (or some other implicit converter) exists in C++.  Without that, you have to expose any non-constructor implicit conversion operators from A -> B as B constructors.

Now maybe this isn't important enough to worry about.  We could say that implicitly_convertible<A,B> means something like this:
- if B is not a pybind11-registered type: A must be a registered type, and must be implicitly convertible (in C++) to B.  Any attempt to pass an A instance as an argument that doesn't accept A instances will try to do the implicit conversion, and only fail with an argument error if there is no such conversion.
- otherwise the (registered) B type will be created as needed when passing some unsupported type Z to a function, but a (declared) B(Z) constructor exists.

As for exposing conversion operators, that's probably not too difficult anyway: something like ```.def("__init__", [](const A &a) { a.operator B(); })``` ought to accomplish that.

wjakob commented on include/pybind11/cast.h:
```diff
@@ -321,6 +344,7 @@ struct type_caster<
     bool load(handle src, bool) {
         py_type py_value;
 
+        // First we try to convert the value directly to a T:
```
is this left over from a previous patch?

jagerman replied:
Whoops, yup.  Fixed.

wjakob commented on include/pybind11/cast.h:
```diff
@@ -310,9 +335,7 @@ template <typename type> class type_caster<std::reference_wrapper<type>> : publi
-struct type_caster<
-    T, typename std::enable_if<std::is_integral<T>::value ||
-                               std::is_floating_point<T>::value>::type> {
+struct type_caster<T, typename std::enable_if<std::is_arithmetic<T>::value>::type> {
```
Why remove the ``is_floating_point``?

jagerman replied:
It wasn't removed, just changed to is_arithmetic, which is true for both integral and floating point types.

wjakob replied:
ok, sounds good!

wjakob commented:
This patch increases the size of generated example module by ~17% (when excluding test case 18, which you added). I've been working very hard to shave off fractions like these in the last releases, so this concerns me somewhat.. 😰 

I wonder if there could be a simple way to reduce the overhead?

I'm assuming that the problem was caused by moving the conversion code the ``std::tuple<>`` type caster. This made the patch less intrusive and easier to understand. The downside is that the tuple template is instantiated for literally every function definition, so those few lines you added translate in quite a bit of extra object code for every function call (vs type, of which there are many fewer).

From a software maintainer perspective, the previous approach was less nice because every ``type_caster`` declarations had to be modified to support implicit conversions.

Just a thought: how about augmenting PYBIND11_TYPE_CASTER so that it inserts a separate function into every type caster (e.g. load_or_convert) which falls back to an implicit conversion upon failure?

Generally the more you can do in templates that are instantiated fewer times the better, and ideally most of the code is in functions that are not templated at all.

Thoughts?

jagerman commented:
Some of that increase is coming from making type_caster_generic::load virtual, which was needed to abstract the try_implicit_conversion (since it needs to call load, but needs to go to the derived instance).  This is only called in two places, however, so simply inlining that bit of code and de-virtualizing load() (in c1a9ecd) reduces the example so size for me by about 5.5%.  (I see a total increase of around 25%, including the example18).

I'll investigate some more changes (including the TYPE_CASTER change) to see what else can be done here.

jagerman commented:
I implemented your suggested change (moving everything to the individual type_casters, mostly picked up via PYBIND11_TYPE_CASTER, with a load_or_convert in type_caster), but the change in so size was very minimal: it reduces the so size increase to 13.4% (over current master) from 14.3% (both counted with example18 disabled), and with the eliminated virtual I pushed in c1a9ecd.

I've commited the change in 408eb61502fe77f4f64c7f2f50ef6640aa36f08a (not currently pushed to this PR) if you want to take a look, because I'm not sure it's worthwhile.

A slight savings, certainly, but the implementation is certainly less elegant, and more intrusive (it's definitely heading back in the direction of the original implementation).  It's also a bit more fragile: for example, subclasses of PYBIND11_TYPE_CASTER-using classes with their own load() method need to provide their own load_or_convert method, so as to avoid the parent's load_or_convert calling the parent's load().

wjakob commented:
A question about your new patch: why not use the same storage object for the implicitly converted object? A boolean flag could be used to indicate whether an implicit conversion took place (i.e. for lifetime management). This would simplify all the ``operator type()`` functions quite a bit I think.

jagerman commented:
See the latest (d2b6719).  This disables implicit casting entirely if a new "implicit.h" header hasn't been included, keeping the .so size unchanged when code doesn't need it, while still providing the feature when desired.

jagerman commented:
As for the other patch: PYBIND11_TYPE_CASTER sets up a Type value, if Type is trivially destructible, we could stash the implicit converted value there with an equivalent of new (&value) OutputType(...).  If not trivially destructible, though, that's not safe--so either we'd have to have `Type *value` for everything, or a mix depending on whether Type is trivially destructible.

Either way, I don't think it's going to save a whole lot--I can't see eliminating one pointer per type_caster instantiation cutting down the .so size a lot.  I think the approach I just pushed is nicer: it's cleaner (general type_casters don't need any support code at all, everything is done by the tuple type_caster), and it only increases the object size for code that actually wants to use the feature.

dean0x7d commented:
I haven't been following the code here closely, but the last change regarding `implicit.h` looks like an ODR violation if the header is included in some translation units but not others. E.g. there would be two `type_caster<Ts...>` classes with identical names and identical type parameters but different definitions depending on the inclusion of `implicit.h` in a translation unit. To avoid undefined behavior, the user would need to ensure to either include the header everywhere or nowhere, but that's an easy mistake to make.

If it should be optional, perhaps a preprocessor define would be a better choice, since it's much easier to enforce for all translation units.

jagerman commented:
Ah yes, you're right, that's not going to work.

jagerman commented:
Any suggestions for how to do it via a macro that wouldn't have exactly the same issue?

jagerman commented:
Oh, I suppose you mean a project-level build macro, rather than a per-translation unit macro.

dean0x7d commented:
Yeah, exactly that. 

However, you should probably wait for @wjakob to comment on this since adding a project-level optional feature may be an additional maintenance burden going forward -- it could potentially double the test suite.

For the feature itself, my guess is that the current implementation can stay mostly intact, just with everything from `implicit.h` moved into `cast.h` and placed under an `#ifdef` guard. A project-level define would ensure the same code is included everywhere. Of course, this may also make it possible to simplify the implementation compared to the separate header, but I haven't dug deep enough into the code to make any meaningful comments.

wjakob commented:
Thanks for chiming in @dean0x7d !

@jagerman: A few more thoughts (sorry to drag this on, but I don't think it's quite ready yet): I think the approach via a separate header file is actually quite nice, though the ABI issue is of course a blocker.

The templated way of potentially adding the extra member and enabling/disabling functions in the tuple caster seems somewhat overkill to me. How about something much more old school? If implicit.h is included, ``#define PYBIND11_IMPLICIT_CONVERSIONS 1`` (set to ``0`` by default, e.g. in ``common.h``).

Next, add a constant to the template parameters of the tuple caster: ``int ImplicitConversions = PYBIND11_IMPLICIT_CONVERSIONS``. This ensures that the types are different (i.e. no ABI problem).

Finally, a higher level remark: the terminology of just calling this an implicit conversion in the code is a bit confusing, as there is already a kind of implicit conversion (and e.g. all the ``type_caster::load`` functions have a template parameter ``convert`` which enables/disables just that.)

jagerman commented:
On another note, your suggestion for the load-or-convert branch (reusing the value storage with a bool for lifetime management) actually makes a surprisingly large difference (e9d7b75): the example .so increase is down to around 9% from ~ 13.4% with that applied just to the type_caster_base (since it already works by pointer rather than value).

Is that an acceptable increase, or do you prefer the opt-in approach?

Re: naming, how about changing "implicit_converted" and the like to "implicit_cpp_instances", and similarly for implicit_casters -> implicit_cpp_casters.

jagerman commented:
The reason for the template approach for enable/disabling rather than the old-school approach is that it isn't affected by header include order.  With the #define, this won't work:

#include <pybind11/pybind11.h>
#include <pybind11/implicit.h>

which seems undesirable.

jagerman commented:
I think that b0202c1 addresses the ODR violation, by defining two type_caster<tuple> instances, enabling based on whether "implicit.h" is included, as before.

As far as I can tell, it isn't feasible to do this with a default template parameter as @wjakob suggested, because type_caster<tuple> is itself already a partial template specialization, which doesn't accept allow template parameters.  It also can't be done via a common base class, because that wouldn't solve the issue: the derived type_caster<tuple> would still be ODR violating, since it is has a different parent in different translation units.  So the approach here: stick all the common class definitions in a macro, with two classes using the type_caster SFINAE parameter to enable one or the other based on whether implicit.h has been included.

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -1064,24 +1064,68 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
+                // conversion value and moves/copies the converted value into it, returning a pointer.
+                // Returns nullptr if input is nullptr.  The caller is responsible for destruction.
+                if (!input) return nullptr;
+                OutputType t = *reinterpret_cast<InputType*>(input);
```
Why the extra move -- isn't this enough? ``return new OutputType(*reinterpret_cast<InputType*>(input));``

jagerman replied:
I started with it that way earlier, but it was erroneous in that it will invoke *explicit* conversions as well, and that could give a different conversion path than you would get in regular C++ code.

For example, in the example classes in example18.cpp, Ex18_B is implicitly convertible to Ex18_E uniquely, via its `operator Ex18_E()`.  Without the construct-and-move approach, you (can) also get the temporary Ex18_E via either `Ex18_E(b.operator double())`, or `Ex18_A(static_cast<A&>(b))`, but those shouldn't be considered for implicit conversion (and aren't, in C++) because `Ex18_E(const double&)` and `Ex18_E(const Ex18_A&)` are both marked explicit.

jagerman replied:
7835704 adds a comment about this, because it is definitely not immediately obvious.

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -1064,24 +1064,68 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
+    auto *output_type = detail::get_type_info(typeid(OutputType));
+    if (output_type) {
+        // Implicit conversion to a registered type, via registered constructors
+        if (!output_type)
```
superfluous if statement.

wjakob replied:
I suppose this should still throw an exception, but probably in the else{} branch below if implicit conversions were not requested by the user.

jagerman replied:
Fixed in f2ad0c8 (removes the if) and afc9800 (throws an error about needing implicit.h when applicable).

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -1064,24 +1064,68 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
+    // Some validity checks on output type:
+    if (!std::is_destructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not destructible");
+    if (!std::is_move_constructible<OutputType>::value && !std::is_copy_constructible<OutputType>::value)
```
Can likely remove ``is_move_constructible`` test here.

jagerman replied:
It's possible for OutputType to be movable but not copyable, but still support implicit conversion (see example below).  We prefer a move constructor, but if that isn't available, fallback to a copy constructor--so this is just checking that at least one of those is available.  This if statements (along with the one before it) should also be compiled away entirely by any decent compiler, since the ::value's are constexpr bools.



Example: implicit conversion of a movable, non-copyable type:
```
class In {};
class Out {
public:
    Out() = default;
    Out(const Out &) = delete;
    Out(Out &&) = default;
    Out(const In&) {}
};

void out(const Out&) { std::cout << "out called\n"; }

int main() {
    In in;
    out(in);
}
```

wjakob commented on include/pybind11/cast.h:
```diff
@@ -735,12 +822,14 @@ template <typename type, typename holder_type> class type_caster_holder : public
-            for (auto &converter : typeinfo->implicit_conversions) {
+            // Walk through the implicit python conversions looking for one that works.
+            for (auto &converter : typeinfo->implicit_python_conversions) {
                 temp = object(converter(src.ptr(), typeinfo->type), false);
```
This is super minor: I would prefer it if the fields were called ``implicit_conversions_python`` and ``implicit_conversions_cpp``

jagerman replied:
Done in 9ab3a96.

wjakob commented:
@jagerman: I think 9% is still a lot to pay for a feature which will not be relevant to most users of this libraries. I know that 9% doesn't sound like a lot -- but look at huge binding projects to realize what this means just in terms of tens of megabytes of object code: https://github.com/pybind/pybind11/commit/86d825f3302701d81414ddd3d38bcd09433076bc#commitcomment-17741296  So if this is the best that can be done, then that's fine -- however, the feature will have to be optional in this case.

One aspect that I don't yet understand is with regards to the lifetime of created objects.
There is naturally no issue with ints, floats and other trivial types. However, what happens when some instance of class "A" is implicitly converted to an instance of class "B"? As far as I understand, your code will delete the "B" instance after the function call.

However, the Python callee could certainly expect to store one of the function parameters in a variable (e.g. class attribute) and expect it to be alive later on. This was never a problem with the previous set of implicit conversions because they are always tracked by an underlying ref-counted PyObject subclass. 

Or am I just missing something?

jagerman commented:
I may well be missing something here, but my thinking on this is that it shouldn't be an issue: the implicitly converted instances are only going to be created immediately before a C++ function/lambda call, and will be destroyed right after the call, i.e. basically what happens for implicit conversion in C++ already.  None of this code applies when calling a python function, however--when the callee is a python function, we're going down a whole other code path, with conversion from C++ types into pybind11-wrapped types, but the implicit C++ conversion code doesn't apply in that case--this is really only for cases when the target class *isn't* pybind-registered, so that that conversion isn't feasible anyway).  If the C++ function that is called with the implicitly converted C++ instances itself calls back to a Python function, I still don't think there's any issue--this is no different than any C++ function with local instances: trying to take references to them is a C++ problem (i.e. not our problem).

That said, I think the current code is slightly too broad with respect to creating the instances--right now I'm pretty sure they can be created by any type_caster<tuple>, but the intention is really to only provide this for an argument-casting tuple.  (I think this would probably lead to a reference error, or at least an unexpected nullptr).  I have code (not quite ready, but will be pushed soon) that addresses this by moving the argument loading bits of type_caster<tuple> into their own subclass, keeping type_caster<tuple> as just a generic tuple caster.

wjakob commented:
ok, sound fair. I guess problems could be expected for functions that take pointer and store it someplace, but that's just asking for trouble and should be fairly obvious to the developer.

jagerman commented:
> The templated way of potentially adding the extra member and enabling/disabling functions in the tuple caster seems somewhat overkill to me. How about something much more old school? If implicit.h is included, #define PYBIND11_IMPLICIT_CONVERSIONS 1 (set to 0 by default, e.g. in common.h).
> 
> Next, add a constant to the template parameters of the tuple caster: int ImplicitConversions = PYBIND11_IMPLICIT_CONVERSIONS. This ensures that the types are different (i.e. no ABI problem).

I've been thinking about solutions to this, and I think it's much trickier than that.  Whether we do the above or use the templated approach, we just back up the ODR problem one step but don't eliminate it--the function initializing the tuple caster would then have different definitions: in one translation unit the local `tuple_caster<std::tuple<...>>` variable is a `tuple_caster<std::tuple<...>, void, 0>` and in the other it's `tuple_caster<std::tuple<...>, void, 1>`.  I explored how far this goes--we'd need to add a similar template parameter to cpp_function::initialize(), which then requires template parameters to the cpp_function constructors, which then requires template parameter for everything in pybind constructing cpp_function (module::def, class_, keep_alive_impl), which then cascades into detail::process_attributes.

And even then, I'm not sure I've tracked everything down.  The point is, it becomes a maintenance nightmare and makes the code more fragile (new could would have to similarly worry about it when calling any of the above) that I'm not even going to try to suggest is worthwhile.  It does have the advantage of allowing you to link implicit-cpp-conversion-enabled code with -disabled code, and lets you enable it via a header include rather than a build time definition.  That could be nice in a large project where implicit conversion is useful in just one part.

So I'll keep plugging away at it to see if I can come up with another approach that minimizes the object size, but if not, I think this will have to work via a compile-time define.

jagerman commented:
This latest commit gets the (non-constant) overhead down to 2.9% on current examples.  (See 6bcf259 commit message for details): basically this eliminates everything from templated code except for two function calls, and accomplishes everything else in non-template code.

I'll squash these commits and investigate the Windows build failure (assuming it's still failing) later tonight.

Comments/feedback?

wjakob commented on docs/advanced.rst:
```diff
@@ -628,10 +629,50 @@ Python side:
+With this statement, our Python code can now call ``func(a)`` and have it
+treated as if we had written ``func(B(a))``.
+
+When including the additional header file :file:`pybind11/implicit.h`, pybind11
```
I guess this is gone :)

jagerman replied:
Whoops, yeah.  Also the example18 reference was wrong, and the class names (A, B) used in the given example were misleading.  Fixed in 6222136.

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -1110,22 +1147,84 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
+    auto &implicit_converters = internals.implicit_conversions_cpp[outidx];
+
+    // If the input type is a subclass of something else in the implicit conversion list, we need to
+    // put this one in just before it (so that this gets found first); otherwise we add to the end
```
Is this comment in sync with the code below? To me, it seems that it it's not a subtype of any element in the list, the while loop will immediately terminate. Afterwards, the element will be inserted at the beginning and not the end of the list.

Related question: is this logic being tested by some part of your tests?

jagerman replied:
Good catch; there should be a ! on the "is subtype" condition.  Added a test and fixed it in c2cc42a.

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -1110,22 +1147,84 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
+NAMESPACE_BEGIN(detail)
+
+template <typename InputType, typename OutputType>
+void* instance_create(void *input) {
```
Propose renaming to: ``implicitly_convert_cpp``

wjakob replied:
Or maybe: ``implicit_conversion_cpp_create`` (and then ``implicit_conversion_cpp_destroy`` below)

jagerman replied:
Done in 1728fab.

wjakob commented on include/pybind11/cast.h:
```diff
@@ -139,6 +153,95 @@ inline PyThreadState *get_thread_state_unchecked() {
+            }
+        }
+        // If the i'th argument didn't load normally, and we didn't find a constructor, we can't do implicit conversion
+        if (constructors.empty() || constructors.back().first != i)
```
Simpler logic: add ``bool found=false;`` above for loop over ``imp_conv``, set to true if found,  and check for it here.

jagerman replied:
Done in 666952b.

wjakob commented on include/pybind11/common.h:
```diff
@@ -260,13 +261,18 @@ struct overload_hash {
     std::unordered_map<const void *, void*> registered_instances;    // void * -> PyObject*
     std::unordered_set<std::pair<const PyObject *, const char *>, overload_hash> inactive_overload_cache;
     std::forward_list<void (*) (std::exception_ptr)> registered_exception_translators;
+    std::unordered_map<std::type_index, std::vector<std::pair<PyTypeObject *, void *(*)(void *)>>> implicit_conversions_cpp;
```
It would be nicer to have single unordered_map here, whose value is a ``struct`` with both the implicit conversion constructor and the destructor.

jagerman replied:
Done in 1fa55c0.

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -1110,22 +1147,84 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
+// the C++ types aren't implicitly convertible.
+template <typename InputType, typename OutputType>
+typename std::enable_if<std::is_convertible<InputType, OutputType>::value>::type
+implicitly_cpp_convertible(detail::type_info &input_type) {
```
propose renaming to ``implicitly_convertible_cpp``

jagerman replied:
1728fab.

wjakob commented on include/pybind11/cast.h:
```diff
@@ -139,6 +153,95 @@ inline PyThreadState *get_thread_state_unchecked() {
 #endif
 }
 
+PYBIND11_NOINLINE inline bool try_implicit_conversion(const bool already_loaded[], const std::vector<PyObject *> &py_args) {
```
Wouldn't it make more sense to use a ``PyObject *`` here, which stores a tuple (instead of std::vector<PyObject *>).  ``PyTuple_SETITEM`` generates just as much object code as std::vector operations, and then the resulting list of arguments is already in the right format to be used by the function call.

wjakob commented:
Hello Jason,

thank you for looking into this -- 2.9% starts to sound like a more reasonable number. I did a quick pass over your code and think that most changes are quite reasonable (e.g. recording the type in all cases when ``descr<>`` is constructed).

One part gives me a bit of a headache though, and that's the stack of implicit conversions you added. The two things that bother me about it are:

1. This is a massive complication of the function dispatcher and other parts of the library. Only the dispatcher ever pushes stuff on it, so I wonder: why can't the normal stack be used for this via standard recursive function calls? (i.e. a local variable in the dispatcher)

2. This is state that is very much global -- not just within a module but also across different Python threads and completely different extension modules built with pybind11. The ``internals`` data structure is shared by all of them, and the implications of this are not clear to me.

Thus it might be good to consider if some other (simpler) designs are feasible here.

Thanks,
Wenzel

wjakob commented:
Here is also something else to ponder:

Since this patch is currently rather invasive (and seems to be getting more so in every iteration :)), I'm also wondering if there are other use cases that might justify merging it. One of them are implicit conversions for C++ multiple inheritance. It could be that your work is already directly applicable to that problem as well, and that would be simply fantastic.

The trouble with multiple inheritance is that an instance of a type ``A`` can't easily be converted to an instance of, say, type ``G`` (with the following classes shown below), because the implicit ``this`` pointer needs to be shifted around while traversing the graph of classes. C++ does this automatically because it knows about the involved types, but pybind11 has no such feature at the moment.

```
A : public B, C { };
B : public D, E { };
C : public F, G { };

class G { void f(); };
```

My thought was that the pybind11 ``class_<>`` setup code for a polymorphic class with multiple inheritance could register a CPP implicit conversion for all of its base classes, i.e.:

```
register_implicitly_convertible<A, B>();
register_implicitly_convertible<A, C>();
register_implicitly_convertible<B, D>();
register_implicitly_convertible<B, E>();
register_implicitly_convertible<C, F>();
register_implicitly_convertible<C, G>();
```

When a method ``G::f()`` is called with an ``A`` instance, my hope was that the recursive implicit conversion code would cause these conversions to take place by running the conversion from ``A`` to ``C`` and then from ``C`` to ``G`` (it would be no problem if that happens in a sort of brute force manner, which tries out some invalid conversions as well)

What do you think? (Note that this is not a request for you to go and implement multiple inheritance ;) but rather a question on whether this PR could be a part of a bigger project)

jagerman commented:
> One part gives me a bit of a headache though, and that's the stack of implicit conversions you added. The two things that bother me about it are:
>
> 1. This is a massive complication of the function dispatcher and other parts of the library. Only the dispatcher ever pushes stuff on it, so I wonder: why can't the normal stack be used for this via standard recursive function calls? (i.e. a local variable in the dispatcher)
> 2.  This is state that is very much global -- not just within a module but also across different Python threads and completely different extension modules built with pybind11. The internals data structure is shared by all of them, and the implications of this are not clear to me.

It seemed to be slightly more efficient to use the global internals stack to avoid needing to pass arguments through the templated code.  It's a very small change though (I estimate around half a percent increase on the per-signature costs, while adding roughly 8KB of constant code size).  But as you point out, it's also more complicated, and potentially a problem with multiple threads interacting, so I've added a commit that uses a new function internals struct, and just passes that through.  (The "loaded" state can also live in the struct).

> 3.  A bit more minor: it would probably make sense to use Python tuples for argument values rather than switching to a C++ vector data structure.
Yes, I switched it back to use the tuple.  I had thought PyTuple_GET_ITEM was actually a function call, but it's just a macro and indeed doesn't change the object size.

wjakob commented on include/pybind11/cast.h:
```diff
@@ -74,6 +75,19 @@ PYBIND11_NOINLINE inline internals &get_internals() {
 
+// Walks up the inheritance tree of `derived` until a registered (python) type is found, and returns
+// that base class type_info.  Returns nullptr if no registered base class is found.
+PYBIND11_NOINLINE inline detail::type_info* get_parent_type_info(PyTypeObject *type) {
```
Looks like this function is not used anywhere (?)

jagerman replied:
Indeed, it's a leftover function from a previous implementation detail.  Removed.

wjakob commented on include/pybind11/common.h:
```diff
@@ -73,6 +73,7 @@
 #include <stdexcept>
 #include <unordered_set>
 #include <unordered_map>
+#include <stack>
```
not needed anymore

jagerman replied:
Gone.

jagerman commented:
Re: multiple inheritance

It might be feasible, and I renamed `try_implicit_conversion` to `loaded_or_alternatives` with this in mind.  I'm not entirely sure if overloading implicitly_convertible() for this is the right way to go, however: base class use is simply a cast, not a conversion, but that's a minor detail.  Whatever the user-facing interface, loaded_or_alternatives() would be the place for it: if regular loading fails, this could then try base class casting for known base classes, then if that fails, implicit conversion (or implicit conversion could be tried first; it probably makes no difference).

One comment regarding:
> When a method G::f() is called with an A instance, my hope was that the recursive implicit conversion code would cause these conversions to take place by running the conversion from A to C and then from C to G (it would be no problem if that happens in a sort of brute force manner, which tries out some invalid conversions as well)
We really don't want to (and I think, cannot) use the implicit conversion mechanism for this for at least a couple reasons:
1. If `G::f()` is called with `this` set to a pointer to a new (temporary) `G` instance, i.e. via implicit conversion, things are not going to work as expected because the G instance *isn't* the base instance of the caller, it's a copy of the caller's `G` superclass instance.
2. `G` might not be implicitly convertible at all, e.g. if it's an abstract base class that *has* `f()` but only has `virtual void f2() = 0;`
3. Even if `G` is not abstract, overridden virtual methods would break: `void f() { g(); }` where g() is virtual but overridden by `C` or `A` won't do the right thing.

But still, this seems quite feasibly by using a similar implementation, but by doing a static_cast<Base*>() pointer cast instead of an implicit conversion.

wjakob commented:
> But still, this seems quite feasibly by using a similar implementation, but by doing a static_cast() pointer cast instead of an implicit conversion.

Exactly, that was my thinking. Ignoring the differences between implicit conversions and static casts (a kind of implicit conversion in a loose sense I suppose), I am wondering:

With the above set of registered implicit conversions (A-> B, A->C, etc..), what would happen if when calling G::f(A()) with your current traversal of the implicit converters?

jagerman commented:
It would not chain (following the C++ implicit conversion rules, at least for non-primitive types).

You *could* however get it working (ignoring the base instance copy issue) if you defined all the possible conversions, that is, if you added not just derived-to-parent but also derived-to-base conversions.

i.e. you'd need `py::implicitly_convertible<A, G>()` as well.

wjakob commented:
Ok, good to know! (those conversions could be automatically added when registering types with multiple inheritance)

One more question: is there ever a situation when your patch causes a chaining behavior? I thought I remembered something like that when looking at your testcases, though perhaps I misunderstood.

wjakob commented:
Looks like MSVC has a problem with the latest version (lots of errors of the type "parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax")

jagerman commented:
Whoops, that's not right.

jagerman commented:
> One more question: is there ever a situation when your patch causes a chaining behavior? I thought I remembered something like that when looking at your testcases, though perhaps I misunderstood.

No, chaining won't happen--the code only looks at implicit conversions that have been registered with the argument type as the output type.  So `py::i_c<A,B>(); py::i_c<B,C>()` won't consider the A->B conversion when looking for a way to convert the arguments to an argument of type C.  If would have to be a bit more complicated to allow that to also detect and avoid chaining loops, but it would also be implementing a C++-level implicit conversion beyond what C++ actually implements.  There's also a second issue with chaining in that we don't allow implicit conversion between two unregistered types (e.g. A->B, B->C, with both B and C unregistered).

So I suppose that, in order to use this mechanism for multiple inheritance, it would involve checking in implicitly_convertible<A,B> whether std::is_base_of<B, A> and, if so, storing a function that simply does a static_cast<A*>(b) (rather than constructing a `new A` from the B).  It wouldn't need the constructor or destructor, but we could still store the base class pointer in implicit_instances, with a flag to say "don't delete this" during `function_call_internals` destruction.

jagerman commented:
See 44f20dc (not currently pushed to this PR branch) for an implementation of the base class casting and chaining.  It changes `py::implicitly_convertible<A,B>()` to do base class casting, rather than creating a new instances, when B is a subclass of A.  Multiple inheritance isn't a problem (it'll evaluate all possible chains that its been told about the first time the conversion from F to T is needed, and then cache the resulting chain (or lack thereof)).

wjakob commented:
Hello Jason,

this patch looks very promising! 
It looks like this could be merged with another patch (https://github.com/BorisSchaeling/pybind11/commit/81586221a8dacbf718991c2488d20b6ef84e29b0) which introduces a nice API for declaring multiple inheritance but does not do the casting/chaining correctly.

I will need to set aside a bit more time to go through it in detail -- probably I will not get to it until next week.

Best,
Wenzel

jagerman commented:
Squashed into one.

wjakob commented:
FYI: I'm traveling this next week, so this PR will remain open for a bit. (I need to sit down for a day or so to merge this and the multiple inheritance change and make sure that I don't totally mess things up ;))

wjakob commented:
I just tried to give this a test drive but ran into a bunch of compilation issues (possibly related to C++14 mode?)
[compiler-errors.txt](https://github.com/pybind/pybind11/files/380166/compiler-errors.txt)

jagerman commented:
Does 0c3925e fix it?

wjakob commented:
Hmm, looks like this got broken now by some of the other PRs -- sorry :(

jagerman commented:
Rebased and squashed.

wjakob commented on include/pybind11/descr.h:
```diff
@@ -27,6 +27,18 @@ NAMESPACE_BEGIN(detail)
+#define PYBIND11_DESCR_TYPE_INFO '~'
+// Character that signals that the type name should be looked up from its type_info
+#define PYBIND11_DESCR_AUTO_TYPE_NAME '%'
+// The above will be ordered in the order listed above, i.e. ${~%} indicates varaible with a type_info
```
typo: variable

wjakob commented:
Hi Jason,

I took a full pass over this PR just now (which has been in the works for quite a while at this point).

Having seen it again in its entirety, I'm sorry to say that it is still too intrusive of a change for me.  When compiling with clang, this commit increases the generated machine code size by 4.2%, which is not even the main problem: it's the serious amount of code infrastructure & heap-allocated data structures that need to be dropped into the hottest path of pybind11, i.e. the dispatcher. This
seems quite counter to recent efforts of shaving unnecessary nanoseconds from the function dispatcher. The other side of this is that it makes the library harder to understand (even after looking at this patch in many iterations, I am still not confident that I understand all the corner cases that led to the current design. It's really the sheer size of the change that bothers me, so there is no single part that I could point to and say "fix this, and it's mergeable" (which I realize is not so helpful). I was more optimistic at earlier points (sorry!), but this was while only looking at individual portions of the PR.

At this point, I'm not so sure how to proceed. The options I see are to simply close the issue or fundamentally rethink the approach. Perhaps there could be a "lite" version of this patch that addresses an important use case in a minimally intrusive way.

Best,
Wenzel

jagerman commented:
Hi Wenzel,

That's a fair assessment, I'll close the PR.

I'll leave [the branch](https://github.com/jagerman/pybind11/tree/implicit-cpp-conversion) around, in case anyone ever wants to revisit this (or something similar) in the future, but I'm okay with dropping it.

Incidentally, one of the comments that has come up in this and several other issues/PRs relates to multiple inheritance.  Just a thought: it would be nice to have an open "collector" issue to capture the bits and pieces that relate and could potentially contribute to this, such as the experimental upcast walker I played with as a tangent of this PR (44f20dc), and (BorisSchaeling@8158622), and could serve as a long-term discussion area for multiple inheritance support.
